Improve news and events + fix some bugs on library advanced search

avocadodev123 opened the pull request:
### **_What are the changes?_**

**_- Things added:_**

1. Separate the news and events tab on the navigation into news and blogs (combine news and blogs) on 1 tab and events on a different tab.
2. For the news&blogs tab I reuse the page for news_list and I create the same events_list page for the events tab.
3. Added default image for news and blogs and events in case the users didn't upload them.
4. Added the search functionalities for both of them (I have tested this).
5. Added send email notifications to Simron when someone subscribed or unsubscribed the newsletter. 
6. Added the remove subscribers/users in controlpanel/teams to delete the users.

**_- Bug fixes:_**
1. Bug fixes for the library advanced search (after using advanced search to search, the users will be able to edit the advanced search options to search again).

**### _How to test?_**
1. Checkout delete user/subscribers functionality in the controlpanel/teams/
2. Checkout news & blogs and events on the navigation.
3. Test subsequent advanced search after using advanced search.

Hi Paul, these are the changes that I made and Simron and Albert had reviewed and agreed on this, I have test the functionality but if you spot any bugs or inappropriate code styles then don't hesitate to ping me.
Again, thank you so much for your help!

paulhoekman reviewed (changes requested):
See code comments - haven't tested it yet but have a look at code comments first and will test after those are processed. Thanks Kent.

paulhoekman commented on src/templates/events.list.html:
```diff
@@ -0,0 +1,166 @@
+
+{% block content %}
+
+<h2 style="text-align: center;">Events</h2>
```
Use a bootstrap class instead of inline css pls.

paulhoekman commented on src/core/migrations/0114_news_type.py:
```diff
@@ -0,0 +1,18 @@
+            model_name='news',
+            name='type',
+            field=models.CharField(choices=[('news', 'News'), ('blog', 'Blog')], default='news', max_length=10),
+        ),
```
You have two migrations with number 0114, following 0113. This is going to create conflicts when running migrations. Please resolve the duplication.

paulhoekman commented on src/community/views.py:
```diff
@@ -17,13 +17,6 @@
 from django import forms
-from django_recaptcha.fields import ReCaptchaField
-from django_recaptcha.widgets import ReCaptchaV2Checkbox
-from django_ratelimit.decorators import ratelimit
```
Are the packages still used somewhere? If not, please remove from requirements file as well.

avocadodev123 replied:
I think I removed them cuz I don't see them in my current requirements.pip file.

paulhoekman commented on src/core/static/css/main.css:
```diff
@@ -1961,6 +1946,9 @@ body.library.search .container.main-content {
   width: 100% !important;
   max-width: unset !important; }
 
+body.library.search .advanced-options {
```
Kent you might want to review your specificity. It is generally very high, with no apparent benefit. https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_cascade/Specificity

I assume this could also be written as `.advanced-options` for instance. Other example:
    `body.library.search .advanced-options .row-4 #boolean-span {`

IDs are unique so you would expect that `#boolean-span` works equally well. Is there a reason for the high specificity used in CSS?

paulhoekman commented on src/core/static/img/blog-default.jpg:
Just double checking these images are released under an open license and don't need crediting?

avocadodev123 replied:
Thank you for this notice Paul! I double checked and it seems like the blog and events images are licensed however the news are not since I got them from this site: https://www.freepik.com/ 
So I changed the blog and events to use the images from this site also.
I could be wrong though so please give me a heads up about this if you notice something.

Thank you so much!

paulhoekman replied:
Copy that Kent, thanks for changing that. I would say let's add a "data-source" attribute to those images, where you list the source URL. That way I can have a look to confirm the other ones, and we will then also have it documented in the code for the future.

paulhoekman commented on src/core/views.py:
```diff
@@ -2151,9 +2202,19 @@ def newsletter(request):
             is_subscribed.delete()
+
+            # Send the email to the admin user
+            send_mail(
```
For this to work remember we need an email service set up; see my mail. In terms of code it's fine, but we just have to set up the mailing service.

paulhoekman commented on src/templates/news.list.html:
```diff
@@ -69,3 +109,61 @@ <h5 class="mb-0">
+
+{% block footer %}
+<!-- Include jQuery -->
+<script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
```
jquery is included in base, shouldn't be necessary to include again

paulhoekman reviewed (changes requested):
Added some more comments in the code after reviewing it on my local machine.

paulhoekman commented on src/templates/library/index.html:
```diff
@@ -5,6 +5,9 @@
+{% endblock %}
+
 {% block css %}
   <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/select2@4.0.13/dist/css/select2.min.css"/>
```
Hi Kent, I noticed that in this file you are manually including various CSS files including main.css and select2 CSS. Please study _base.html, which is always used as a baseline for every file. Please note:
- main.css is always included in each file; no need to load separately
- you can pass `"load_select2¨: True` to context and it will load the select2 files

Please also review other files, I saw that you have included main.css manually a few times. Thanks.

paulhoekman replied:
Hi Kent,
Please see:
https://github.com/Metabolism-of-Islands/metabolism-of-islands-platform/blob/master/src/templates/_base.html#L40
We used SCSS before but I have changed that to CSS to make sure your direct changes to the css file apply. But let me merge this and I'll then remove the line for ease of use.

avocadodev123 commented:
Hi @paulhoekman . Again, thank you so much for these thoroughly review, I really appreciate your help on this. 
Sorry for the late fix as I was off on the weekend (usually). 
I have fixed most of the reviews and some comments that are not clear I had put the questions for them so please review that.
Other than that, I have tested the functionalities of these changes again after the fix and they seem to still work fine. Please let me know if anything comes to your mind! I really appreciate it!

paulhoekman commented on src/core/views.py:
```diff
@@ -447,21 +447,39 @@ def news_events_list(request, header_subtitle=None):
+def events_list(request, header_subtitle=None):
+
+    list = Event.objects.all().distinct() # no need filtering as the only the MOI events are in the database
+
```
Hey Kent, I used "list" myself in the past to indicate there is a list of items, but it was actually a poor choice as there is already a list() function in python. Whenever you write new code, simply use the plural name of what you are querying, or something like that. In this case, use "events" for instance.

paulhoekman commented on src/core/views.py:
```diff
@@ -931,14 +949,30 @@ def controlpanel(request, space=None):
     if not has_permission(request, request.project, ["curator", "admin", "publisher"]):
         unauthorized_access(request)
+
+    users = RecordRelationship.objects.filter(record_child_id=request.project).exclude(relationship_id=13)
```
You can remove `.exclude(relationship_id=13)` because that was another one of those things related to the multi-project setup. No longer a concern for MoI. You can also remove the comment on line 959. And is there a reason you are moving this to a separate variable instead of straight in the `context`? I don't mind, but just wondering about the motivation.

paulhoekman commented on src/core/views.py:
```diff
@@ -931,14 +949,30 @@ def controlpanel(request, space=None):
+        try:
+            user = users.get(id=user_id)
+            # Perform the deletion
+            user.delete()
```
To stick to the default structure used elsewhere, please move this up, above the `context`. Unless there is a specific need, the structure of each block is: 

- Variables/queries/logic
- Context
- Return render...

Let's stick to that unless there is a reason to change it.

paulhoekman commented on src/templates/contribution/notifications.html:
```diff
@@ -30,7 +30,12 @@ <h1>Notifications</h1>
         </td>
         <td>
-          {{ each.record.message.date_created|naturaltime }}<br>
+          {% if each.record.message.date_created|naturaltime %}
```
I would start the condition like so:

          {% if each.record.message.date_created %}

paulhoekman commented on src/templates/library/index.html:
```diff
@@ -150,19 +138,19 @@
+    .advanced-options .row-4 #boolean-span.changing,
+    .advanced-options .row-4 #field-span.changing,
+    .advanced-options .row-4 #contains-span.changing,
+    .advanced-options .row-4 #term-span.changing {
```
Just double checking: why is ` .advanced-options .row-4` needed? An ID should be unique and normally has the highest specificity, unless we defined something conflicting before (which we shouldn't, so we should then fix that one I'd say). Can you check on this?

paulhoekman commented on src/templates/library/index.html:
```diff
@@ -150,19 +138,19 @@
+    .advanced-options .row-4 #boolean-span.changing::after,
+    .advanced-options .row-4 #field-span.changing::after,
+    .advanced-options .row-4 #contains-span.changing::after,
+    .advanced-options .row-4 #term-span.changing::after {
```
Same here

paulhoekman commented on src/templates/news.list.html:
```diff
@@ -1,44 +1,74 @@
           {% if each.image %}
             <img src="{{ each.image.url }}" class="card-img-top" alt="">
+          {% else %}
+            {% if each.get_article_type_display|upper == "NEWS" %}
```
Let's simplify, no need to convert things in the condition:

            {% if each.article_type == "news" %}

paulhoekman commented on src/templates/news.list.html:
```diff
@@ -1,44 +1,74 @@
             <img src="{{ each.image.url }}" class="card-img-top" alt="">
+          {% else %}
+            {% if each.get_article_type_display|upper == "NEWS" %}
+              <img src="{% static "img/news-default.jpg" %}" class="card-img-top" alt="">
```
This is where it would be useful to add the `data-source` attribute.

paulhoekman commented on src/templates/news.list.html:
```diff
@@ -1,44 +1,74 @@
             <h5 class="card-title">
               {{ each }}
+              <br>
               <small class="d-inline-block text-muted">{{ each.date }}</small>
```
Date shouldn't be part of the `<h5>` 
Please move outside of it. Thanks.

paulhoekman commented on src/templates/newsletter.html:
```diff
@@ -36,6 +36,12 @@ <h1>Newsletter: {{ PROJECT }}</h1>
+      <div class="form-group">
+        <label for="institution">Institution</label>
+        {% if not request.user.is_authenticated %}
+          <input type="institution" class="form-control" id="institution" name="institution">
```
How exactly does this work? If the user not not logged in, there is an institution field? But if the user is logged in, there is a label but not an input?

avocadodev123 replied:
Sorry for this one, when the user is logged in, there should not be a label as well. Thanks for this. I just fixed it.

paulhoekman commented on src/templates/library/index.html:
```diff
@@ -740,7 +728,16 @@ <h3>Publications</h3>
 <script src="https://cdn.jsdelivr.net/npm/select2@4.0.13/dist/js/select2.min.js"></script>
+{% if request.GET.before or request.GET.after or request.GET.type or request.GET.space or request.GET.terms %}
+  <script>
+    $(document).ready(function() {
```
Also standardize use of quotation marks: double ones please.

paulhoekman commented on src/templates/library/index.html:
```diff
@@ -774,9 +771,9 @@ <h3>Publications</h3>
-    const $bar = $('body.library.search .expand-advanced-search-bar');
-    const $options = $('body.library.search .advanced-options');
-    const $select = $('body.library.search .search-select');
+    const $bar = $('.expand-advanced-search-bar');
```
Same here - Also standardize use of quotation marks: double ones please.

paulhoekman reviewed (changes requested):
Thanks for the updates Kent. I followed up on your comments and I added a few more! Almost there, thanks for hanging in there.

avocadodev123 commented:
> Thanks for the updates Kent. I followed up on your comments and I added a few more! Almost there, thanks for hanging in there.

Hi Paul, I made some changes based on your requests, please let me know if anything still need to be changed. I also added the data-source attribute into some elements and remove the load main.css static file on some files that already has base.html extended. Thank you very much!

paulhoekman commented:
Thanks Kent. Will review tomorrow in detail. For now: please add the full URL of each image in the data-source. We need to know exactly who authored it, what the license is, etc and that requires more than just the website. Pls add the source link for each image. Thanks.

paulhoekman commented on src/templates/newsletter.html:
```diff
@@ -36,6 +36,12 @@ <h1>Newsletter: {{ PROJECT }}</h1>
         {% endif %}
       </div>
+      <div class="form-group">
+        {% if not request.user.is_authenticated %}
```
Please put this outside of the <div>, otherwise you have an empty <div> for logged-in users.

paulhoekman commented on src/core/views.py:
```diff
@@ -2151,24 +2181,47 @@ def newsletter(request):
+            
+            send_mail(
+                subject="Newsletter Subscription",
+                message=f"User with an email {email} from institution {institution} has subscribed to a newsletter.",
```
If a user is logged in, won't it read ... `from institution None has subscribed`... ?

avocadodev123 replied:
Hi Paul, yeah, Simron interested in to know the institution like where they are from if the user is not logged in but subscribe to the newsletter. If the user is authenticated then no need to do so.

I can remove the instituition part if the user is logged when sending the email.

paulhoekman replied:
Noted Kent. And yes sounds good -- I understand the motivation and it makes sense, but from a coding viewpoint it's most appropriate to not include "institution None" in the subject I'd say. Small detail but it's the kind of stuff that keeps the code clean and sensible for future use.

paulhoekman commented on src/templates/news.list.html:
```diff
@@ -1,44 +1,73 @@
           {% if each.image %}
             <img src="{{ each.image.url }}" class="card-img-top" alt="">
+          {% else %}
+            {% if each.get_article_type_display == "News" %}
```
See my previous comment around this condition. Here you are checking the "display" value but that can change. Simply validate against the database value.

paulhoekman commented on src/templates/news.list.html:
```diff
@@ -47,19 +76,28 @@ <h5 class="card-title">
             {% if each.image %}
               <img class="rounded" src="{{ each.image.url }}" alt="">
+            {% else %}
+              {% if each.get_article_type_display == "News" %}
```
Also update this condition please

paulhoekman commented on src/templates/events.list.html:
```diff
@@ -0,0 +1,166 @@
+          {% if each.image %}
+            <img src="{{ each.image.url }}" class="card-img-top" alt="">
+          {% else %}
+            <img src="{% static "img/events-default.jpg" %}" class="card-img-top" alt="" data-source="https://www.freepik.com/free-vector/event-manager-typographic-header-celebration-meeting-organization-planning-pr-company-business-mass-media-social-network-promotion-isolated-vector-illustration_27953002.htm#fromView=search&page=2&position=1&uuid=eee7b969-61a6-43aa-a342-306bb0608d65&query=event+ ...
```
I thought I typed this before but I don't see that my comments were saved. Perhaps a duplicate, but if not: please attribute the images properly. These specific images you chose require attribution (see source). Either add attribution or select images that do not require so. Thanks.

avocadodev123 replied:
Hi Paul, I prompted ChatGPT for the new images so I don't think we need the attribution anymore. Thanks!

avocadodev123 commented:
Hi Paul, I have fixed the latest comments from you. Please let me know if there is anything else need to be accounted for. 

Again, much appreciate your help on this!

avocadodev123 commented:
Hi @paulhoekman , I don't know if this will be available to be on the prototype by tomorrow. 
It's because Simron wants to let his students to test the website during this weekend, however most of the bugs fixes are on this update and I am not sure if it will be resolved after this is merged into the website as well. Thus I want to have sometimes to go over to test them again before letting the users to test them.

Moreover, I tested the "remove the bookmark item" on the prototype website and somehow it returns 400 Bad Request, but I tested on my local and it doesn't behave like that, hence, I cannot determine what's going wrong. Do you mind help me to test this and pull the log to see why it returns 400 Bad Request when you remove the bookmarked item?

Thank you very much!!!

paulhoekman commented on src/templates/_base.html:
```diff
@@ -75,6 +75,10 @@
   {% block head %}{% endblock %}
 
+  {% if menu == "library" %}
+    <link href="{% static 'css/main.css' %}?rf12" rel="stylesheet" type="text/css" />
```
Hi Kent, I will fix this after merging, but just FYI: the main.css inclusion should not be conditional; it should happen on every page. I will remove the condition.

avocadodev123 replied:
Hi Paul, can you investigate on this? If you take a look on the library page -> advanced search options, if I include the main.css file below the include load select 2, I would have the style of the main.css overwrite the load select2 which is what I want, that's why I brute forced if the menu is library I want my main.css to overwrite the style of the load select2.

paulhoekman replied:
Hi Kent, 
Can you help me by giving me:
- An example of a specific CSS style that clashes? One is enough.
- An overview on which pages this is used on
I will then investigate. Thanks

avocadodev123 replied:
Hi Paul, i have attached the images below:

1. This is what happening right now on the website (the main.css is above the load select 2 in the _base.html):
<img width="1247" alt="image" src="https://github.com/user-attachments/assets/07da3f7d-ad8a-4506-b702-06449a91d1e5" />

2. This is when I move the main.css file to below the load select2 on the _base.html:
<img width="1256" alt="image" src="https://github.com/user-attachments/assets/a4d58c41-b9ac-42d9-bd5c-3a90898d17cd" />

Because on the main.css I have an overwriting select2 block: Please look from line 740-866 on the main.css file. So, I guess the main.css should be put below the load select2 to overwrite some select2 styles right? Please correct me if I am wrong cuz i am not a very good CSS guy.

The pages is just the Library Page then you click the advanced search options, and on the code it is the index.html in the library templates folder.

Thanks so much Paul!

avocadodev123 replied:
Ohhhh okay I see how it is. I tested it and it worked now! Thanks Paul! Can you push it to the development?

paulhoekman replied:
OK great! It's on prototype. Cheers.
